core/trino-main/src/main/java/io/trino/operator/output/columnar/IsolatedPositionsAppenderFactory.java

```diff
@@ -0,0 +1,285 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.output.columnar;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import io.airlift.bytecode.DynamicClassLoader;
+import io.trino.collect.cache.NonEvictableLoadingCache;
+import io.trino.operator.output.columnar.PositionsAppender.TypedPositionsAppender;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.type.AbstractIntType;
+import io.trino.spi.type.AbstractLongType;
+import io.trino.spi.type.BooleanType;
+import io.trino.spi.type.CharType;
+import io.trino.spi.type.DecimalType;
+import io.trino.spi.type.DoubleType;
+import io.trino.spi.type.SmallintType;
+import io.trino.spi.type.TinyintType;
+import io.trino.spi.type.Type;
+import io.trino.spi.type.UuidType;
+import io.trino.spi.type.VarbinaryType;
+import io.trino.spi.type.VarcharType;
+import io.trino.sql.gen.IsolatedClass;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+
+import java.util.Objects;
+
+import static io.airlift.slice.SizeOf.SIZE_OF_LONG;
+import static io.trino.collect.cache.SafeCaches.buildNonEvictableCache;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Isolates the {@code PositionsAppender} class per type and block tuples.
+ * Type specific {@code PositionsAppender} implementations manually inline {@code Type#appendTo} method inside the loop
+ * to avoid virtual(mega-morphic) calls and force jit to inline the {@code Block} and {@code BlockBuilder} methods.
+ */
+public class IsolatedPositionsAppenderFactory
+        implements PositionsAppenderFactory
+{
+    private final NonEvictableLoadingCache<CacheKey, PositionsAppender> cache;
+
+    public IsolatedPositionsAppenderFactory()
+    {
+        this.cache = buildNonEvictableCache(
+                CacheBuilder.newBuilder().maximumSize(1000),
+                CacheLoader.from(key -> createAppender(key.type)));
+    }
+
+    @Override
+    public PositionsAppender create(Type type, Class<? extends Block> blockClass)
+    {
+        return cache.getUnchecked(new CacheKey(type, blockClass));
+    }
+
+    private PositionsAppender createAppender(Type type)
+    {
+        Class<? extends PositionsAppender> appenderClass = findDedicatedAppenderClassFor(type);
+
+        if (appenderClass == null) {
+            return new TypedPositionsAppender(type);
+        }
+
+        return isolateAppender(appenderClass);
+    }
+
+    private Class<? extends PositionsAppender> findDedicatedAppenderClassFor(Type type)
+    {
+        if (type instanceof AbstractLongType) {
```

Comment:
I think this is wrong. You can check for `FixedWithType` and use the size to know if it's 1/2/4/8/16. This will remove many cases below, and handle `IpAddressType` that is missed. I think all of the `SlicePositionsAppender` checks can be replaced with `VariableWidthType`.